plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OracleClientModule.java

```diff
@@ -53,33 +54,75 @@ public void configure(Binder binder)
-        connectionProperties.setProperty(OracleConnection.CONNECTION_PROPERTY_INCLUDE_SYNONYMS, String.valueOf(oracleConfig.isSynonymsEnabled()));
-        connectionProperties.setProperty(OracleConnection.CONNECTION_PROPERTY_REPORT_REMARKS, String.valueOf(oracleConfig.isRemarksReportingEnabled()));
-
         if (oracleConfig.isConnectionPoolEnabled()) {
-            return new OraclePoolConnectionFactory(
-                    config.getConnectionUrl(),
-                    connectionProperties,
-                    credentialProvider,
-                    oracleConfig.getConnectionPoolMinSize(),
-                    oracleConfig.getConnectionPoolMaxSize(),
-                    oracleConfig.getInactiveConnectionTimeout());
+            return oraclePoolConnectionFactory;
+        }
+        else {
+            return retryingConnectionFactory;
         }
+    }
 
-        return new RetryingConnectionFactory(new DriverConnectionFactory(
-                new OracleDriver(),
+    @Provides
+    @Singleton
+    public OraclePoolConnectionFactory getOraclePoolConnectionFactory(
+            BaseJdbcConfig config,
+            CredentialProvider credentialProvider,
+            OracleConfig oracleConfig,
+            @Named("oracleConnectionProperties") Properties connectionProperties)
+            throws SQLException
+    {
+        return new OraclePoolConnectionFactory(
                 config.getConnectionUrl(),
                 connectionProperties,
                 credentialProvider,
-                openTelemetry));
+                oracleConfig.getConnectionPoolMinSize(),
+                oracleConfig.getConnectionPoolMaxSize(),
+                oracleConfig.getInactiveConnectionTimeout());
+    }
+
+    @Provides
+    @Singleton
+    @Named("retryingDelegate")
+    public ConnectionFactory getDriverConnectionFactory(
+            OracleDriver oracleDriver,
+            BaseJdbcConfig config,
+            CredentialProvider credentialProvider,
+            OpenTelemetry openTelemetry,
+            @Named("oracleConnectionProperties") Properties connectionProperties)
+    {
+        return new DriverConnectionFactory(
+                oracleDriver,
+                config.getConnectionUrl(),
+                connectionProperties,
+                credentialProvider,
+                openTelemetry);
+    }
+
+    @Provides
+    @Singleton
+    @Named("oracleConnectionProperties")
+    public Properties getOracleConnectionProperties(OracleConfig oracleConfig)
+    {
+        Properties connectionProperties = new Properties();
+        connectionProperties.setProperty(OracleConnection.CONNECTION_PROPERTY_INCLUDE_SYNONYMS, String.valueOf(oracleConfig.isSynonymsEnabled()));
+        connectionProperties.setProperty(OracleConnection.CONNECTION_PROPERTY_REPORT_REMARKS, String.valueOf(oracleConfig.isRemarksReportingEnabled()));
+        return connectionProperties;
+    }
+
+    @Provides
+    public OracleDriver getOracleDriver()
```

Comment:
What is the point of adding `OracleDriver` to guice context? It is only used to create to create `DriverConnectionFactory`.

Reply:
The drive behind this is NOT to mix DI with no-DI. Now hidden this as implementation detail behind PrivateModule. Hope this addresses your issues.